### ISSUE TITLE
Support combinations of precision plugins

### ISSUE THREAD

**Issue by carmocca** [Lightning-AI/lightning#18679]
### Description & Motivation

Both the Fabric and Trainer strategies are designed to have a single plugin enabled from the beginning to the end of the program.

This has been fine historically, however, some strategy implementations define their own plugin that accomplish a specific task that migth be supported by a different plugin, but with different implementation details.

For instance, single device training with `bf16-true` precision will use the [`HalfPrecision` plugin](https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/precision/half.py#L39-L52)
But FSDP training with `bf16-true` precision will use the [`FSDPPrecision` plugin](https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/precision/fsdp.py#L80-L121)

Recent cutting-edge plugins such as `TransformerEnginePrecision` and `BitsandbytesPrecision also use some of the `bf16-true` functionality ([example](https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/precision/transformer_engine.py#L93-L125)). So there's a lot of overlap.

The challenge then becomes: how do I enable TransformerEngine to work with FSDP if they both want to control the plugin interface?

Note that I'm using these specific classes to show the point, but the design issue applies to any strategy that wants a specific plugin class. DeepSpeed and XLA are also examples of this.

### Pitch

Continuing the example, there's 3 ways this could be solved:
1. The naive way: Create a `TransformerEngineFSDPPrecision`. This is simple and effective but it creates maintainability problem.
2. The independent way: If there are no dependencies between the plugins, we could support `plugins=[TransformerEnginePrecision(), FSDPPrecision()]`
3. The smart way: Create an abstraction that is able to compose two (or more?) plugins together and is itself a plugin. There's some [precedent for this with the `CheckpointIO` plugins](https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/precision/transformer_engine.py#L93-L125).

**Comment by naveenkumarmarri** [comment]
@carmocca is there a way to use `TransformerEnginePrecision` with FSDP?

**Comment by carmocca** [comment]
Not for now: https://github.com/NVIDIA/TransformerEngine/issues/401

**Comment by naveenkumarmarri** [comment]
@carmocca to understand it better, where is the bottleneck? does transformer engine has to support it or FSDP has to support this?

**Comment by carmocca** [comment]
TransformerEngine, and then we wouldn need to integrate whatever is changes into Lightning.

cc @sbhavani in case you know about the progress for this

**Comment by naveenkumarmarri** [comment]
@sbhavani is there a timeline that you’re targeting  for the feature to be
available in FSDP?

**Comment by naveenkumarmarri** [comment]
@carmocca any plan to support FP8 training for deepspeed strategy?

**Comment by carmocca** [comment]
@sbhavani I see https://github.com/NVIDIA/TransformerEngine/tree/main/examples/pytorch/fsdp exists now. Is your last comment still valid?

**Comment by sbhavani** [comment]
@carmocca  Yes upstream support for FP8 storage and comms. in PyT's FSDP implementation is still in progress. However, with lazy weight initialization in that example, you can use FP8 with FSDP.

**Comment by function2-llx** [comment]
I would like to suggest another use case that two different precision plugins may be needed.
Suppose my lightning module is composed of a transformer and a CNN. Is it possible to use bf16 for the transformer and 16-mixed for the CNN?

**Comment by sbhavani** [comment]
@carmocca it might be worth adding FSDP+FP8 support via TE. HF Accelerate just added FSDP+FP8 support as a reference: https://github.com/huggingface/accelerate/blob/main/benchmarks/fp8/fsdp.py

**Comment by psr-ai** [comment]
@sbhavani if I get it right, as of today, the support for FP8 + FSDP is not available with PyTorch Lightning?

**Comment by lantiga** [comment]
Correct. It's very much on our radar though. Anyone wants to run ahead and submit a PR?

I do think that right now:
> Create a TransformerEngineFSDPPrecision. This is simple and effective but it creates maintainability problem.

is going to have the highest ROI. Once we have this fully working we can think about generalizing in the future.

Here's the starting point if anyone would like to give it a shot:
https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/fabric/plugins/precision/transformer_engine.py

**Comment by radulescupetru** [comment]
@lantiga Any update on this? Is this also the reason why `ModelParallelStrategy` does not work with `precision=transformer-engine`?
